execution_engine/src/core/engine_state/mod.rs

```diff
@@ -1798,4 +1801,205 @@ where
+            correlation_id,
+            auction_contract.contract_wasm_hash(),
+            self.config.use_system_contracts(),
+            &preprocessor,
+        ) {
+            Ok(module) => module,
+            Err(_) => {
+                return Ok(StepResult::PreconditionError);
+            }
+        };
+
+        if !self.system_contract_cache.has(auction_hash) {
+            self.system_contract_cache
+                .insert(auction_hash, auction_module.clone());
+        }
+
+        let system_account = Account::new(
+            SYSTEM_ACCOUNT_ADDR,
+            Default::default(),
+            URef::new(Default::default(), AccessRights::READ_ADD_WRITE),
+            Default::default(),
+            Default::default(),
+        );
+        let authorization_keys = {
+            let mut ret = BTreeSet::new();
+            ret.insert(SYSTEM_ACCOUNT_ADDR);
+            ret
+        };
+        let mut named_keys = auction_contract.named_keys().to_owned();
+        let gas_limit = Gas::new(U512::from(std::u64::MAX));
+        let deploy_hash = {
+            // seeds address generator w/ protocol version
+            let bytes: Vec<u8> = step_request.protocol_version.value().into_bytes()?.to_vec();
+            Blake2bHash::new(&bytes).value()
+        };
+
+        let base_key = Key::from(protocol_data.auction());
+
+        let slashed_validators = match step_request.slashed_validators() {
+            Ok(slashed_validators) => slashed_validators,
+            Err(error) => {
+                error!(
+                    "failed to deserialize validator_ids for slashing: {}",
+                    error.to_string()
+                );
+                return Ok(StepResult::Serialization(error));
+            }
+        };
+
+        let slash_args = runtime_args! {"validator_public_keys" => slashed_validators};
+
+        let (_, execution_result): (Option<()>, ExecutionResult) = executor.exec_system_contract(
+            DirectSystemContractCall::Slash,
+            auction_module.clone(),
+            slash_args,
+            &mut named_keys,
+            Default::default(),
+            base_key,
+            &system_account,
+            authorization_keys.clone(),
+            BlockTime::default(),
+            deploy_hash,
+            gas_limit,
+            step_request.protocol_version,
+            correlation_id,
+            Rc::clone(&tracking_copy),
+            Phase::Session,
+            protocol_data,
+            SystemContractCache::clone(&self.system_contract_cache),
+        );
+
+        if execution_result.has_precondition_failure() {
+            return Ok(StepResult::PreconditionError);
+        }
+
+        let distribute_args = runtime_args! {}; // TODO: distribute args?
+
+        let (_, execution_result): (Option<()>, ExecutionResult) = executor.exec_system_contract(
+            DirectSystemContractCall::DistributeRewards,
```

Comment:
I thought that we were supposed to `distribute` **after** `run_auction`...

Reply:
did this